rfc/parameterizedMetrics.md

```diff
@@ -0,0 +1,124 @@
+Parameterized Metric Support
+============================
+
+Parameterized metrics are a subset of dynamic request metrics.  Specifically they focus on modifying the calculation of metrics by altering values used in the calculation of existing metrics or mapping their values at aggregation time.
+
+The expected contract of parameterized metrics is as follows: Given a base metric with api name `metricName`, it can be suffixed with a procedure-call list of keyword parameter names and values.  These values will be used in the calculation of the resulting new metric.
+
+So
+```
+"metricName(parameterName=parameterValue)"
+```
+
+will result in a dynamic metric based on `metricName`, but substituting the parameter value when calculating it.  Also, the above string will be used as the column name in the resulting report response schema. 
+
+Motivation
+----------
+
+Standard metrics in Fili are configured at load time.  However if reporting requires many different similar caluclations, such as adjusting precision or converting values of different units into a shared aggregate, or applying custom metric filters, it may be impractical or even impossible to configure all the interesting report metrics.
+
+With parameterized metrics, open-ended variations may be supported for a given conceptual metric calculation, simplifying configuration and possibly unlocking otherwise impractical freedom in metric calculation.
+
+
+The structure of a parameterized metric:
+----------------------------------------
+
+```
+metricName(parameterName=value)
+```
+
+metricName -> an api metric name which should appear as a resource at the `/v1/metrics/metricName` endpoint.
+
+Balanced outer parentheses are the signal that a metric is to be processed as parameterized metric.  Imbalanced parentheses or unescaped internal parentheses will probably result in request errors.
+
+parameterName -> an identifier for a parameter supported by this metric.  Identifier names should follow the same format rules as api resource identifiers.
+
+value -> a value to be used for creating a modified version of baseMetric.  URL encoding may be required if some of the characters in the value contain special characters.  Escaping values containing ampersand is mandatory and it may also be necessary to encode values containing comma or parentheses depending on implementation.
+
+If multiple parameter names are used for this dimension they should be joined by commas with no intermediate spaces (similar to the filter value syntax)
+
+i.e.
+
+```
+metricName(parameterName=value,otherParameter=otherValue)
```

Comment:
How will duplicate metrics with the same param values be handled?

Reply:
I would expect it to be handled exactly the same as a query requesting "pageviews,pageviews". If that doesn't error, neither should this. A more interesting question is "pageviews(as=pv1),pageviews(as=pv2)", which I think should be successful, even though its the same value being generated.

Reply:
I am fine with returning `400` as it does today:

`Duplicate metric(s) are not allowed in the API request..`

Let's just call it out that the same is true with param values.

Reply:
I hope this RFC can focus on the parameterized metric spec.  Details for `as`, `filter(s)` and other built in params can be their own RFC.

Reply:
I think these cases need to be considered when evaluating parameterized metrics.
If we define a syntax that doesn't handle the very next two things we'll want to do with it, it's not a good spec.